### PR TITLE
Fujitsu T111 keyboard

### PR DESCRIPTION
**Pull request by DmNosachev**
New keyboard: Fujitsu N860-2500-T111
QMK-conversion by directly connecting to the existing matrix 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/handwired/t111/keymaps/oleg/keymap.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+#include "debug.h"
+#include "version.h"
+
+enum custom_keycodes {
+  VRSN,
```
Comment (drashna) [inline keyboards/handwired/t111/keymaps/oleg/keymap.c:21]: You need/want to set this to `SAFE_RANGE`, or you'll run into weird issues, later on. 

```suggestion
  VRSN = SAFE_RANGE,
```

File: keyboards/handwired/t111/keymaps/oleg/keymap.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    _______, KC_BTN1, KC_MS_U, KC_BTN2, KC_BTN3, _______, _______, _______, KC_SYSREQ, KC_SLCK, KC_PAUS,        XP(ULSB, ULCB), XP(URSB, URCB),        _______,   _______, _______, _______, _______, _______, _______,
+    _______, KC_MS_L, KC_MS_D, KC_MS_R, KC_BTN5, _______, _______, _______, _______,   _______, XP(USCO, UCLN), XP(UAPO, UQOT),        _______,                          _______, _______, _______, _______, _______, _______,
+    _______,          _______, _______, KC_BTN4, _______, _______, _______, _______,   XP(UCOM, ULTH), XP(UDOT, UGTH), XP(USLS, UQUE),        KC_APP,               _______,    _______, _______, _______, _______, _______, _______,
+                   _______, _______,                               _______,                                     _______,                                   _______, _______,    _______,   _______
+    )
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+	  case VRSN:
```
Comment (drashna) [inline keyboards/handwired/t111/keymaps/oleg/keymap.c:130]:
```suggestion
      case VRSN:
```

File: keyboards/handwired/t111/keymaps/oleg/keymap.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      if (record->event.pressed) {
+        SEND_STRING (QMK_KEYBOARD "/" QMK_KEYMAP " @ " QMK_VERSION);
+      }
+      return false;
+      break;
+	  
+	  case MC1:
```
Comment (drashna) [inline keyboards/handwired/t111/keymaps/oleg/keymap.c:137]:
```suggestion
   case MC1:
```

File: keyboards/handwired/t111/keymaps/oleg/keymap.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      if (record->event.pressed) {
+        send_unicode_string("[]");
```
Comment (drashna) [inline keyboards/handwired/t111/keymaps/oleg/keymap.c:139]: Why not just? 
```suggestion
        SEND_STRING("[]");
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/handwired/t111/keymaps/oleg/keymap.c (hunk 4 above)
Comment (DmNosachev) [reply keyboards/handwired/t111/keymaps/oleg/keymap.c:139]: I've made 'oleg' keymap by request of the keyboards owner. His idea was to have an ability to input certain characters (e.g. this pair of square brackets) independently of current language layout in OS. Russian layout has х and ъ on [ and ].

File: keyboards/handwired/t111/t111.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+|-------------------------------------------------------------------------------------------------------------------------------|     |
+|         | caps| graph |                    Space                      | alt |                | left|right|  |     0     |  .  |     |
+`-------------------------------------------------------------------------------------------------------------------------------------'
+*/
+
+#define LAYOUT( \
+  K30,      K40, K11, K21, K31, K41, K51, K61, K71, K01, K12, K22, K32, K42, K52, K62, K72, \
+  K20, K13, K23, K33, K43, K53, K63, K73, K03, K14, K24, K34, K44, K54, K58,   K64, K74,   K15, K06, K76, K66, \
+  K10,  K17, K27, K37, K47, K57, K67, K77, K07, K18, K28, K38, K48,     K5A,   K68, K78,   K1B, K00, K70, K60, \
+  K0A,   K19, K29, K39, K49, K59, K69, K79, K09, K1A, K2A, K3A, K4A,           K6A, K7A,   K2B, K3B, K4B, K5B, \
```
Comment (noroadsleft) [inline keyboards/handwired/t111/t111.h:43]: QMK custom is that ISO Enter's location is toward the end of the home row.

```suggestion
  K10,  K17, K27, K37, K47, K57, K67, K77, K07, K18, K28, K38, K48,            K68, K78,   K1B, K00, K70, K60, \
  K0A,   K19, K29, K39, K49, K59, K69, K79, K09, K1A, K2A, K3A, K4A,    K5A,   K6A, K7A,   K2B, K3B, K4B, K5B, \
```

File: keyboards/handwired/t111/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * |-------------------------------------------------------------------------------------------------------------------------------|     |
+ * |         | gui |  alt  |                    Space                      | alt |                | left|right|  |     0     |  .  |     |
+ * `-------------------------------------------------------------------------------------------------------------------------------------'
+*/
+    [0] = LAYOUT(
+    KC_NO,            KC_F1,  KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11, KC_F12, KC_NO, KC_PSCR, KC_SLCK, KC_PAUS,
+    KC_ESC,  KC_1,    KC_2,   KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL, KC_GRV, KC_BSPC,   KC_INS,  KC_DEL,   KC_NLCK, KC_PSLS, KC_PAST, KC_PMNS,
+    KC_TAB,  KC_Q,    KC_W,   KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,        KC_ENT,    KC_PGUP, KC_PGDN,  KC_P7,   KC_P8,   KC_P9,   KC_EQL,
+    KC_LCTRL, KC_A,   KC_S,   KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_BSLS,                   KC_HOME, KC_END,   KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
```
Comment (noroadsleft) [inline keyboards/handwired/t111/keymaps/default/keymap.c:38]:
```suggestion
    KC_TAB,  KC_Q,    KC_W,   KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,                   KC_PGUP, KC_PGDN,  KC_P7,   KC_P8,   KC_P9,   KC_EQL,
    KC_LCTRL, KC_A,   KC_S,   KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_BSLS,        KC_ENT,    KC_HOME, KC_END,   KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
```

File: keyboards/handwired/t111/info.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
...16.5, "y":3.5}, {"label":"4", "x":17.75, "y":3.5}, {"label":"5", "x":18.75, "y":3.5}, {"label":"6", "x":19.75, "y":3.5}, {"label":",", "x":20.75, "y":3.5}, {"label":"Shift", "x":0, "y":4.5, "w":2.25}, {"label":"Z", "x":2.25, "y":4.5}, {"label":"X", "x":3.25, "y":4.5}, {"label":"C", "x":4.25, "y":4.5}, {"label":"V", "x":5.25, "y":4.5}, {"label":"B", "x":6.25, "y":4.5}, {"label":"N", "x":7.25, "y":4.5}, {"label":"M", "x":8.25, "y":4.5}, {"label":"<", "x":9.25, "y":4.5}, {"label":">", "x":10.25, "y":4.5}, {"label":"?", "x":11.25, "y":4.5}, {"x":12.25, "y":4.5}, {"label":"Shift", "x":13.25, "y":4.5, "w":2}, {"label":"\u0432\u2020\u2018", "x":15.5, "y":4.5}, {"label":"\u0432\u2020\u201c", "x":16.5, "y":4.5}, {"label":"1", "x":17.75, "y":4.5}, {"label":"2", "x":18.75, "y":4.5}, {"label":"3", "x":19.75, "y":4.5}, {"label":"Enter", "x":20.75, "y":4.5, "h":2}, {"label":"Caps", "x":1.75, "y":5.5}, {"label":"Graph", "x":2.75, "y":5.5, "w":1.5}, {"x":4.25, "y":5.5, "w":8}, {"label":"Alt", "x":12.25, "y":5.5, "w":1.5}, {"label":"\u0432\u2020\u0452", "x":15.5, "y":5.5}, {"label":"\u0432\u2020\u2019", "x":16.5, "y":5.5}, {"label":"0", "x":17.75, "y":5.5, "w":2}, {"label":".", "x":19.75, "y":5.5}]
```
Comment (noroadsleft) [inline keyboards/handwired/t111/info.json:9]: Moving the ISO Enter key's object.

```suggestion
            "layout": [{"label":"Break", "x":0, "y":0}, {"label":"PF1", "x":2, "y":0}, {"label":"PF2", "x":3, "y":0}, {"label":"PF3", "x":4, "y":0}, {"label":"PF4", "x":5, "y":0}, {"label":"PF5", "x":6, "y":0}, {"label":"PF6", "x":7, "y":0}, {"label":"PF7", "x":8, "y":0}, {"label":"PF8", "x":9, "y":0}, {"label":"PF9", "x":10, "y":0}, {"label":"PF10", "x":11, "y":0}, {"label":"PF11", "x":12, "y":0}, {"label":"PF12", "x":13, "y":0}, {"label":"PF13", "x":14, "y":0}, {"label":"PF14", "x":15, "y":0}, {"label":"PF15", "x":16, "y":0}, {"label":"PF16", "x":17, "y":0}, {"label":"Esc", "x":0, "y":1.5}, {"label":"!", "x":1, "y":1.5}, {"label":"@", "x":2, "y":1.5}, {"label":"#", "x":3, "y":1.5}, {"label":"$", "x":4, "y":1.5}, {"label":"%", "x":5, "y":1.5}, {"label":"^", "x":6, "y":1.5}, {"label":"&", "x":7, "y":1.5}, {"label":"*", "x":8, "y":1.5}, {"label":"(", "x":9, "y":1.5}, {"label":")", "x":10, "y":1.5}, {"label":"_", "x":11, "y":1.5}, {"label":"+", "x":12, "y":1.5}, {"label":"~", "x":13, "y":1.5}, {"label":"bkspc", "x":14, "y":1.5, "w":1.25}, {"label":"Ins", "x":15.5, "y":1.5}, {"label":"Cls", "x":16.5, "y":1.5}, {"label":"*", "x":17.75, "y":1.5}, {"label":"/", "x":18.75, "y":1.5}, {"label":"+", "x":19.75, "y":1.5}, {"label":"-", "x":20.75, "y":1.5}, {"label":"Tab", "x":0, "y":2.5, "w":1.5}, {"label":"Q", "x":1.5, "y":2.5}, {"label":"W", "x":2.5, "y":2.5}, {"label":"E", "x":3.5, "y":2.5}, {"label":"R", "x":4.5, "y":2.5}, {"label":"T", "x":5.5, "y":2.5}, {"label":"Y", "x":6.5, "y":2.5}, {"label":"U", "x":7.5, "y":2.5}, {"label":"I", "x":8.5, "y":2.5}, {"label":"O", "x":9.5, "y":2.5}, {"label":"P", "x":10.5, "y":2.5}, {"label":"{", "x":11.5, "y":2.5}, {"label":"}", "x":12.5, "y":2.5}, {"label":"El", "x":15.5, "y":2.5}, {"label":"Dup", "x":16.5, "y":2.5}, {"label":"7", "x":17.75, "y":2.5}, {"label":"8", "x":18.75, "y":2.5}, {"label":"9", "x":19.75, "y":2.5}, {"label":"=", "x":20.75, "y":2.5}, {"label":"Caps Lock", "x":0, "y":3.5, "w":1.75}, {"label":"A", "x":1.75, "y":3.5}, {"label":"S", "x":2.75, "y":3.5}, {"label":"D", "x":3.75, "y":3.5}, {"label":"F", "x":4.75, "y":3.5}, {"label":"G", "x":5.75, "y":3.5}, {"label":"H", "x":6.75, "y":3.5}, {"label":"J", "x":7.75, "y":3.5}, {"label":"K", "x":8.75, "y":3.5}, {"label":"L", "x":9.75, "y":3.5}, {"label":":", "x":10.75, "y":3.5}, {"label":"\"", "x":11.75, "y":3.5}, {"label":"|", "x":12.75, "y":3.5}, {"label":"Enter", "x":13.75, "y":2.5, "w":1.5, "h":2}, {"label":"Del", "x":15.5, "y":3.5}, {"label":"Home", "x":16.5, "y":3.5}, {"label":"4", "x":17.75, "y":3.5}, {"label":"5", "x":18.75, "y":3.5}, {"label":"6", "x":19.75, "y":3.5}, {"label":",", "x":20.75, "y":3.5}, {"label":"Shift", "x":0, "y":4.5, "w":2.25}, {"label":"Z", "x":2.25, "y":4.5}, {"label":"X", "x":3.25, "y":4.5}, {"label":"C", "x":4.25, "y":4.5}, {"label":"V", "x":5.25, "y":4.5}, {"label":"B", "x":6.25, "y":4.5}, {"label":"N", "x":7.25, "y":4.5}, {"label":"M", "x":8.25, "y":4.5}, {"label":"<", "x":9.25, "y":4.5}, {"label":">", "x":10.25, "y":4.5}, {"label":"?", "x":11.25, "y":4.5}, {"x":12.25, "y":4.5}, {"label":"Shift", "x":13.25, "y":4.5, "w":2}, {"label":"\u0432\u2020\u2018", "x":15.5, "y":4.5}, {"label":"\u0432\u2020\u201c", "x":16.5, "y":4.5}, {"label":"1", "x":17.75, "y":4.5}, {"label":"2", "x":18.75, "y":4.5}, {"label":"3", "x":19.75, "y":4.5}, {"label":"Enter", "x":20.75, "y":4.5, "h":2}, {"label":"Caps", "x":1.75, "y":5.5}, {"label":"Graph", "x":2.75, "y":5.5, "w":1.5}, {"x":4.25, "y":5.5, "w":8}, {"label":"Alt", "x":12.25, "y":5.5, "w":1.5}, {"label":"\u0432\u2020\u0452", "x":15.5, "y":5.5}, {"label":"\u0432\u2020\u2019", "x":16.5, "y":5.5}, {"label":"0", "x":17.75, "y":5.5, "w":2}, {"label":".", "x":19.75, "y":5.5}]
```

File: keyboards/handwired/t111/rules.mk [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = yes         # Console for debug
```
Comment (noroadsleft) [inline keyboards/handwired/t111/rules.mk:20]:
```suggestion
CONSOLE_ENABLE = yes        # Console for debug
```

File: keyboards/handwired/t111/rules.mk [hunk 9]
```diff
@@ -0,0 +1,29 @@
+MCU = STM32F103
+
+# GENERIC STM32F103C8T6 board - stm32duino bootloader
+MCU_LDSCRIPT = STM32F103x8_stm32duino_bootloader
+BOARD = STM32_F103_STM32DUINO
+STM32_BOOTLOADER_ADDRESS = 0x80000000
+
+DFU_ARGS = -d 1eaf:0003 -a2 -R
+DFU_SUFFIX_ARGS = -v 1eaf -p 0003
```
Comment (noroadsleft) [inline keyboards/handwired/t111/rules.mk:9]:
```suggestion
DFU_ARGS = -d 1EAF:0003 -a2 -R
DFU_SUFFIX_ARGS = -v 1EAF -p 0003
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/10262/files).

**Comment by noroadsleft** [comment]
It seems the re-request for reviews may have marked my review comments as resolved, but the issues I flagged are still outstanding. Was that deliberate?

**Comment by DmNosachev** [comment]
> It seems the re-request for reviews may have marked my review comments as resolved, but the issues I flagged are still outstanding. Was that deliberate?

Sorry, that's because I don't know much about how GitHub's UI handles pull requests. I have marked all your suggested changes as _Resolved_ and thought that it makes them to be applied automatically. But I still can see _noroadsleft requested changes_ on the bottom of the page and can't find any other buttons and/or links to agree with proposed changes.

**Comment by noroadsleft** [comment]
> Sorry, that's because I don't know much about how GitHub's UI handles pull requests. I have marked all your suggested changes as _Resolved_ and thought that it makes them to be applied automatically. But I still can see _noroadsleft requested changes_ on the bottom of the page and can't find any other buttons and/or links to agree with proposed changes.

Okay, there's a couple different systems at play here:

- **Marking a conversation as resolved**:
  "I have completed the requested action."
- **[user] requested changes**:
  You see this when a reviewer requests changes, up until the point that user changes the status of their request. They may request more changes, or approve the changes you've made.

When you have changes requested, you'll see something like this:

image

You can commit that suggestion by clicking the button, but if you go to the Files Changed tab, you gain access to the **Add suggestion to batch** feature:

image

Instead of committing that change directly, it adds that suggestion to a queue, where all the suggestions you add to the queue can be committed at once. (I sort of selfishly prefer this because it saves CPU cycles on our CI infrastructure.)

I've unresolved my previous review comments. I don't recall if committing the suggestions (individually or in a batch) marks the conversations as resolved automatically or not, but I'll review them anyway when the PR updates.

Hope this helps.

**Comment by noroadsleft** [comment]
[Missed one](https://github.com/qmk/qmk_firmware/pull/10262#discussion_r486988268), but you found the right buttons. :slightly_smiling_face:

**Review by noroadsleft — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**